ouroboros-network/test/Test/PeerState.hs

```diff
@@ -0,0 +1,462 @@
+
+prop_subscriptionWorker
+    :: SocketType
+    -> Int -- local address
+    -> Int -- remote address
+    -> ArbValidPeerState IO ArbDiffTime
+    -> (Fun (ArbDiffTime, Int, ()) (ArbSuspendDecision ArbDiffTime))
+    -> ArbErrorPolicies
+    -> (Blind (ArbApp Int))
+    -> Property
+prop_subscriptionWorker
+    sockType localAddr remoteAddr (ArbValidPeerState ps)
+    returnCallback (ArbErrorPolicies errPolicies)
+    (Blind (ArbApp merr app))
+  =
+    ioProperty $ do
+      doneVar :: StrictTMVar IO () <- newEmptyTMVarM
+      tbl <- newConnectionTable
+      peerStatesVar <- newPeerStatesVar
+      worker nullTracer
+             tbl
+             peerStatesVar
+             (mkSocket sockType remoteAddr)
+             (\ss s -> do
+              s' <- socketStateChangeTx ss s
+              case ss of
+                CreatedSocket{} -> pure s'
+                ClosedSocket{}  -> tryPutTMVar doneVar () >> pure s')
+             completeTx
+             (main doneVar)
+             (Just localAddr)
+             (Just localAddr)
+             (\_ a b -> getFirst (First a <> First b))
+             (const Nothing)
+             (pure $ ipSubscriptionTarget nullTracer peerStatesVar [remoteAddr])
+             1
+             (\sock -> app sock
+                `finally`
+                (void $ atomically $ tryPutTMVar doneVar ()))
+  where
+    ps0 :: PeerState IO (Time IO)
+    ps0 = getArbDiffTime <$> ps
+
+    completeTx = completeApplicationTx
+       nullTracer
+       (ErrorPolicies
+          errPolicies
+          (\t addr r -> fmap getArbDiffTime . getArbSuspendDecision $ case returnCallback of
+              Fn3 f -> f (ArbDiffTime t) addr r
+              _     -> error "impossible happend"))
+
+    main :: StrictTMVar IO () -> Main IO (PeerStates IO Int (Time IO)) Bool
+    main doneVar s = do
+      done <- maybe False (const True) <$> tryReadTMVar doneVar
+      let r = case sockType of
+            WorkingSocket   -> case merr of
+              -- TODO: we don't have access to the time when the transition was
+              -- evaluated.
+              Nothing -> True
+              Just (ArbException e) -> transitionSpec remoteAddr ps0
+                                        (evalErrorPolicies
+                                          (ApplicationException e)
+                                          errPolicies)
+                                        s
+            AllocateError _ -> True
+            ConnectError e  -> transitionSpec remoteAddr ps0
+                                              (evalErrorPolicies
+                                                (ConnectionException e)
+                                                errPolicies)
+                                              s
+      if done
+        then pure r
+        else if r then retry else pure r
+
+
+-- transition spec from a given state to a target states
+transitionSpec :: Ord addr
```

Comment:
transitionSpec seems to have a lot of cases, how about using `tabulate` to get an insight in which paths we tests.

Reply:
Good point, I'll add it.

Reply:
@karknu I actually don't see a way to do that.  `transitionSpec` is called on every transition that happens when the subscription worker is running (which is more than once).  Just adding `tabulate` to it will not work. I will add instead a basic tabulate which just shows what are the inputs of the `prop_subscriptionWorker`.